src/tools/illink/src/linker/Linker.Steps/MarkStep.cs

```diff
@@ -2453,23 +2452,150 @@ void MarkInterfaceImplementations (TypeDefinition type)
 			if (!type.HasInterfaces)
 				return;
 
-			foreach (var iface in type.Interfaces) {
+			var ifaces = GetRecursiveInterfaceImplementations (type);
+			foreach (var (ifaceType, impls) in ifaces) {
 				// Only mark interface implementations of interface types that have been marked.
 				// This enables stripping of interfaces that are never used
-				if (ShouldMarkInterfaceImplementation (type, iface))
-					MarkInterfaceImplementation (iface, new MessageOrigin (type));
+				if (ShouldMarkInterfaceImplementationList (type, impls, ifaceType))
+					MarkInterfaceImplementationList (impls, new MessageOrigin (type));
 			}
 		}
 
-		protected virtual bool ShouldMarkInterfaceImplementation (TypeDefinition type, InterfaceImplementation iface)
+		List<(TypeReference, List<InterfaceImplementation>)> GetRecursiveInterfaceImplementations (TypeDefinition type)
 		{
-			if (Annotations.IsMarked (iface))
+			List<(TypeReference, List<InterfaceImplementation>)> firstImplementationChain = new ();
+
+			AddRecursiveInterfaces (type, [], firstImplementationChain, Context);
+			Debug.Assert (firstImplementationChain.All (kvp => Context.Resolve (kvp.Item1) == Context.Resolve (kvp.Item2.Last ().InterfaceType)));
+
+			return firstImplementationChain;
+
+			static void AddRecursiveInterfaces (TypeReference typeRef, IEnumerable<InterfaceImplementation> pathToType, List<(TypeReference, List<InterfaceImplementation>)> firstImplementationChain, LinkContext Context)
+			{
+				var type = Context.TryResolve (typeRef);
+				if (type is null)
+					return;
+				// GeVt all explicit interfaces of this type
+				foreach (var directIface in type.Interfaces) {
+					//var directlyImplementedType = Context.Resolve (directIface.InterfaceType);
+					var directlyImplementedType = directIface.InterfaceType.TryInflateFrom (typeRef, Context);
+					if (directlyImplementedType is null) {
+						continue;
+					}
+					if (!firstImplementationChain.Any (i => InterfaceTypeEquals(i.Item1, directlyImplementedType, Context))) {
```

Comment:
Not quite sure what you mean. When we mark an interface implementation, we'll always mark the first implementation chain that leads to the interface type, even if another less direct chain is marked.

Reply:
We'd never mark a less direct chain when handling "this type needs to have an implementation of this interface". We may run into a situation where there is a less direct chain that was marked as a result of other dependencies, but if the type needs a chain of implementations to an interface marked, we'll always mark the most direct chain in addition to the existing chain so that method resolution works as expected.